### ISSUE TITLE
EDD_Payment::setup_currency() may return the wrong currency

### ISSUE THREAD

**Issue by daigo75** [easydigitaldownloads/Easy-Digital-Downloads#4162]
Method `EDD_Payment::setup_currency()` is designed to return the active currency when an order (payment) doesn't have any currency associated to it. Such behaviour is incorrect, for a couple of reasons:

- A payment must always have a currency associated to it. The absence of such informatiom should be treated as an error.
- The active currency may change dynamically in multi-currency environments. Returning such value could mean showing a payment total as 123.45 USD/EUR/GBP/CAD, depending on which currency is active at the moment.

Returning the base currency would also be incorrect. As mentioned, an order could have been placed in any of the active currency, and there is no correct "default" that could be applicable. Therefore, an order without a currency should be considered "corrupt".

Note
--
I couldn't determine the circumstances in which the currency is not saved with an order. In any case, it's important to to flag the order as "unusable" if that happens, as there would be no way to manage it correctly.

**Comment by pippinsplugins** [comment]
If we don't have a currency explicitly set in the payment meta, we should default to `edd_get_currency()`.

**Comment by daigo75** [comment]
Unfortunately, that's incorrect. If there is no currency set in the meta, then there is no currency that applies to the order, and no default is valid. There is no guarantee that the order was placed in any specific currency, therefore assuming that the active one is the one to take is wrong.

Besides, unless you set the currency in the meta when it's missing (i.e. you "patch" the missing data and store it in the metadata), retrieving the currency dynamically will only cause issues in a multi-currency environment, as the order currency will change dynamically depending on what the active currency is. That is, if you have an order for 124.56 "no currency", you will see it as 123.45 USD when the active currency is USD, as 123.45 GBP when the currency is GBP, and so on.

Bottom line: there is no valid default.

**Comment by cklosowski** [comment]
@daigo75 there needs to be some sort of default used or else we'll have some issues with gateways. In all cases prior to 2.5, we've always defaulted the gateway on a payment of the store default, this defaulting is not new.

**Comment by cklosowski** [comment]
Here is where the code was previously run:
https://github.com/easydigitaldownloads/Easy-Digital-Downloads/blob/release/2.4/includes/payments/functions.php#L1120

The `EDD_Payment` class has just centralized all of the payment helper functions into a unified class. We've not changed any functionality on the default currency of a payment when missing payment meta, however we should do an `empty` check instead of an `isset` in case it is empty.

**Comment by daigo75** [comment]
@cklosowski there is a difference between the currency used before payment and the one used after. When a payment is created the currency **must** be stored in its metadata (in the database), in that moment. You can't take the active currency at a later stage, as it will change dynamically.

My point is that there cannot be a payment without currency in the metadata, as that will cause all sorts of issues. As soon as a payment is added to the database, the currency must be part of its meta (before it's processed any further).

**Comment by pippinsplugins** [comment]
@daigo75 We don't disagree. A currency code should always be stored, but we also need to properly fallback if one wasn't stored for some reason.

**Comment by daigo75** [comment]
@pippinsplugins and that's the issue: there is no valid fallback, as I already explained.

**Comment by pippinsplugins** [comment]
@daigo75 Yep, got it. We're fixing it.

**Comment by daigo75** [comment]
If you review an order and you see it as 123.45 USD, then you open it later and you see it as 123.45 GBP, I'm quite sure that you will start wondering what's going on. This is the type of issue that has to be avoided, or shop owners will get alarmed.

My opinion is that this issue can be prevented quite easily by forcing a currency whenever it's missing. At a glance, the best place to do that could be `EDD_Payment::insert_payment()`. If the currency is missing at that stage, no matter the reason, then you can call `edd_currency()` to get the active one. Such currency will then be "set in stone" against the payment, and passed to the payment gateways, to the reports and to all the nice pages that EDD provides. This should guarantee that a payment always has a currency, and the fallback in getting the currency will no longer be needed. :+1:

**Comment by daigo75** [comment]
I should be able to send you a pull request with a proposed patch, hang on.

**Comment by pippinsplugins** [comment]
We do need to ensure it's always inserted with a currency but we also need to ensure that when one isn't (due to technical error most likely), that it displays something.

**Comment by daigo75** [comment]
@pippinsplugins that's the key: in a system with six currencies, which one is the correct one?

**Comment by pippinsplugins** [comment]
@daigo75 The store default. Always.

**Comment by daigo75** [comment]
In which case, you should not call `edd_get_currency()`, which returns the active currency (whose value changes dynamically), but `edd_get_option( 'currency' )`. And, in such case, you could patch" the order and store the missing currency against it, instead of determining it dynamically.

**Comment by pippinsplugins** [comment]
No, `edd_get_currency()` should always be used, never `edd_get_option( 'currency' )`. Using the helper function ensures the proper filters are run on it.

**Comment by daigo75** [comment]
That's precisely the cause of the issue. With `edd_get_currency()` you don't get the shop's base currency, but the **active** currency (the two match, by coincidence, only when you have a single currency in the shop). That function can return **any** of the available currencies, at any given time, and this causes the issue I reported in the first place: the order with the missing currency will appear different depending on which currency is active at the moment.

**Comment by pippinsplugins** [comment]
`edd_get_currency()` returns the shop default. Your plugin modifies the response to return the "active" currency. That means you need to modify how your filter works to respect the shop default as well.

**Comment by daigo75** [comment]
I'm afraid that's not possible. The filter is not aware that is should return shop's base currency, because it cannot be: there are no elements to determine that context (i.e. there is nothing that says "now we really need the shop's default").

**Comment by pippinsplugins** [comment]
As long as payment records don't get created without a currency, this should never be a problem. Have you seen payments created without a currency recently? If so, do you see it when running EDD core by itself without extensions?

**Comment by daigo75** [comment]
I've seen some payment created on a customer's system before they used my plugin, but I can't determine the reason why that happened. All I see is that those payments are all "pending".

**Comment by pippinsplugins** [comment]
If they are payments from 2+ years ago, that's very possible.

**Comment by daigo75** [comment]
I created a PR that should help preventing the issue. If you have some time to spare, you can have a look at that (I just wrote it).

I was also thinking that, as a fallback, we could have a filter for the default payment currency too. That would not change the logic as it is, and it would provide a context in which the `edd_get_currency()` could return shop's base currency. Example:

~~~
/**
 * Setup the currency code
 *
 * @since  2.5
 * @param  array $payment_meta The payment meta
 * @return string              The currency for the payment
 */
private function setup_currency() {
	$currency = isset( $this->payment_meta['currency'] ) ? $this->payment_meta['currency'] : apply_filters( 'edd_default_payment_currency', edd_get_currency(), $this );
	return $currency;
}
~~~

**Comment by daigo75** [comment]
Since the issue only affects multi-currency environments, the filter would allow the plugins that implement such feature to take action as they need. A call to that filter would automatically imply that the payment has no currency associated to it (if I'm not mistaken, the ternary operator is short-circuited), and the user could be informed about it. The multi-currency solution could then provide an interface to allow the shop owner to rectify the order, by manually choosing the currency.

That would cover all scenarios, without changing any of the existing logic.

**Comment by pippinsplugins** [comment]
The PR looks good. We'll test and review for 2.5.5.

**Comment by daigo75** [comment]
Should I also make a PR with the new filter for payments' default currency? It will take me a minute.

By the way, while we are at it: why is the `EDD_Payment` class declared as `final`? It could be useful to be able to subclass it.

**Comment by pippinsplugins** [comment]
@daigo75 Another PR would be fine though that should be a separate issue/PR.

The class is marked as `final` for now since it is brand new. Once it is fully stable and we're ready for developers to start extending it, we will remove the `final`. See this post for more details: https://easydigitaldownloads.com/development/2015/12/15/introducing-edd_payment/

**Comment by daigo75** [comment]
Ok, I will open a new issue/feature request and submit a PR.

As for the class, it's a welcome addition. I had created my own Order class (just for consistency with other plugins, as a payment is the equivalent of an order in other ecommerce systems) about two years ago, because there wasn't one available. It does almost the same things that your EDD class I will be happy to move to a subclass of EDD_Payment once it becomes stable. :+1:

**Comment by daigo75** [comment]
Done: #4165.

**Comment by pippinsplugins** [comment]
:+1:
